docs/spec/spec.md

```diff
@@ -0,0 +1,1821 @@
+10. [Protocol-upgrade](#10-protocol-upgrade)
+    * 10.1. [HTTP2](#101-http2)
+        * 10.1.1. [Push promise and promise response](#1011-push-promise-and-promise-response)
+    * 10.2. [Websocket](#102-websocket)
+
+## 1. Overview
+This specification elaborates on Basic Auth authentication and authorization for all the Ballerina listeners and
+clients. The HTTP, gRPC, GraphQL, WebSocket, WebSub protocol-based listeners and clients are secured according to this
+specification.
+
+The ballerina language provides the first class support for writing network oriented programming models in a more 
+initiative manner. The ballerina-http standard library consumes the rich language construct and creates the 
+programming model to write services.
+
+## 2. Components
+### 2.1. Listener
+The HTTP listener object receives network data from a remote process according to the HTTP transport protocol and 
+translates the received data into invocations on the resources functions of services that have been 
+attached to the listener object. The listener provides the interface between network and services.
+
+As defined in [Ballerina 2021R1 Section 5.7.4](https://ballerina.io/spec/lang/2021R1/#section_5.7.4) the Listener has 
+the object constructor and life cycle methods such as attach(), detach(), 'start(), gracefulStop(), and immediateStop().
+
+#### 2.1.1. Auto binding
+If a service is attached to the listener, then the listener starts listening on the given port after executing 
+attach() and start() methods. HTTP listener can be declared as follows honoring to the generic 
+[listener declaration](https://ballerina.io/spec/lang/2021R1/#section_8.3.1)
+
+```ballerina
+// Listener object constructor
+listener http:Listener serviceListener = new(9090, config = {httpVersion: "2.0"});
+```
+```ballerina
+// Service attaches to the Listener
+service http:Service /sessionTest on serviceListener {
+
+}
+```
+
+#### 2.1.2. Late binding
+
+Users can programmatically start the listener by calling each lifecycle method as follows.
+
+```ballerina
+// Listener object constructor
+listener http:Listener serviceListener = new(9090);
+
+public function main() {
+    error? err1 = serviceListener.attach(s, “/basePath”);
+    error? err2 = serviceListener.start();
+    //...
+    error? err3 = serviceListener.gracefulStop();
+}
+
+http:Service s = service object {
+    resource function get sayHello(http:Caller caller) {}
+}
+```
+
+### 2.2. Service
+Service is a collection of resources functions, which are the network entry points of a ballerina program. 
+In addition to that a service can contain public and private functions which can be accessed by calling with `self`.
+
+#### 2.2.1. Service type
+```ballerina
+public type Service distinct service object {
+
+};
+```
+Above distinct type is provided by HTTP module and user can include the type as `*http:Service` to refer it.
+The comprehensive typing support is yet to be added to the language
```

Comment:
```suggestion
The comprehensive typing support is yet to be added to the language.
```

Reply:
I think better to add a note say until that compiler plugins are used to validate the services.